### PR TITLE
Fixes #463 and addresses #469

### LINKED ISSUE #463: Improve docs with Events timeline image
As there is a lot of issues where folks seem not to understand well ignite's event handling, idea is to improve the docs with a visual representation like [this one](https://github.com/pytorch/ignite/issues/462#issuecomment-475677991). Thoughts ?

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #463 and  addresses #469

Description:
Added Timeline and Events and FAQ

Check list:
* [x] Documentation is updated (if required)

cc @anmolsjoshi

### CODE REVIEW

**Comment by anmolsjoshi** [comment]
Possible things to add:
- Why are there two functions to run evaulator (train and validation sets)? How to correctly set EarlyStopping if running evaluating on both datasets.
- Adding a custom event (TBPTT engine example)
- Calculating metrics during training (why it's not recommended)

**Comment by vfdev-5** [comment]
@anmolsjoshi sounds good !

**Comment by anmolsjoshi** [comment]
@vfdev-5 please review latest commit.

File: docs/source/faq.rst [hunk 1]
```diff
@@ -39,6 +39,52 @@ As a consequence, the following code is correct too:
...
 
+Creating Custom Events based on Forward/Backward Pass
+-----------------------------------------------------
+
+There are cases where the user might want to add events based on the loss calculation and backward pass. Ignite provides
+flexibility to the user to allow for this:
+
+.. code-block:: python
+
+    class BackpropogationEvents(Enum):
```
Comment (vfdev-5) [inline docs/source/faq.rst]: BackpropAgationEvents and same for others

File: docs/source/faq.rst [hunk 2]
```diff
@@ -39,6 +39,52 @@ As a consequence, the following code is correct too:
...
+
+    def update(engine, batch):
+        model.train()
+        opitmizer.zero_grad()
+        x, y = process_batch(batch)
+        y_pred = model(x)
+        loss = loss_fn(y_pred, y)
+        engine.fire_event(BackpropogationEvents.BACKPROPOGATION_STARTED)
+        loss.backward()
+        optimizer.step()
```
Comment (vfdev-5) [inline docs/source/faq.rst]: https://docs.fast.ai/callback.html#Callback.on_backward_end
engine.fire_event(BackpropogationEvents.BACKPROPOGATION_ENDED) should be fired before optimizer.step()

File: docs/source/faq.rst (hunk 1 above)
Comment (anmolsjoshi) [reply docs/source/faq.rst]: To clarify are you asking to edit the name?

Comment (vfdev-5) [reply docs/source/faq.rst]: typo: BackpropogationEvents ->BackpropAgationEvents

**Comment by vfdev-5** [comment]
I left some comments. I would call the event as fast.ai : BACKWARD_BEGIN, BACKWARD_END

File: docs/source/faq.rst [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+.. code-block:: python
+
+    class BackpropagationEvents(Enum):
+        """
+        Events based on Backpropagation
+        """
+        BACKPROPAGATION_STARTED = 'backpropagation_started'
+        BACKPROPAGATION_ENDED = 'backpropagation_ended'
+        OPTIMIZER_STEP_END = 'optimizer_step_end'
```
Comment (vfdev-5) [inline docs/source/faq.rst]:
```suggestion
        OPTIMIZER_STEP_ENDED = 'optimizer_step_ended'
```

File: docs/source/faq.rst [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        model.train()
+        opitmizer.zero_grad()
+        x, y = process_batch(batch)
+        y_pred = model(x)
+        loss = loss_fn(y_pred, y)
+        engine.fire_event(BackpropagationEvents.BACKPROPOGATION_STARTED)
+        loss.backward()
+        engine.fire_event(BackpropagationEvents.BACKPROPOGATION_ENDED)
+        optimizer.step()
+        engine.fire_event(BackpropagationEvents.OPTIMIZER_STEP_END)
```
Comment (vfdev-5) [inline docs/source/faq.rst]:
```suggestion
        engine.fire_event(BackpropagationEvents.OPTIMIZER_STEP_ENDED)
```

File: docs/source/faq.rst [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+flexibility to the user to allow for this:
+
+.. code-block:: python
+
+    class BackpropagationEvents(Enum):
+        """
+        Events based on Backpropagation
+        """
+        BACKWARD_BEGIN = 'backpropagation_started'
+        BACKWARD_END = 'backpropagation_ended'
```
Comment (vfdev-5) [inline docs/source/faq.rst]: Sorry, maybe `BACKWARD_ENDED` is better ?

Comment (anmolsjoshi) [reply docs/source/faq.rst]: I think BACWARD_END would be better. This way a user would decorate as follows:

```python
@trainer.on(BackpropEvents.BACKWARD_END)
def...
```

I think that reads better

File: docs/source/faq.rst (hunk 4 above)
Comment (anmolsjoshi) [reply docs/source/faq.rst]: Although I did incorporate this, I'm rethinking this for this reason - https://github.com/pytorch/ignite/pull/471#discussion_r272749559

File: docs/source/faq.rst (hunk 3 above)
Comment (anmolsjoshi) [reply docs/source/faq.rst]: See here - https://github.com/pytorch/ignite/pull/471#discussion_r272749559

**Comment by vfdev-5** [comment]
Okay, lets define it once for all, otherwise there are too much notions:
```python
    class BackpropEvents(Enum):
        """
        Events based on back propagation
        """
        BACKWARD_STARTED = 'backward_started'
        BACKWARD_COMPLETED = 'backward_completed'
        OPTIM_STEP_COMPLETED = 'optim_step_completed'

    def update(engine, batch):
        model.train()
        opitmizer.zero_grad()
        x, y = process_batch(batch)
        y_pred = model(x)
        loss = loss_fn(y_pred, y)
        engine.fire_event(BackpropEvents.BACKWARD_STARTED)
        loss.backward()
        engine.fire_event(BackpropEvents.BACKWARD_COMPLETED)
        optimizer.step()
        engine.fire_event(BackpropEvents.OPTIM_STEP_COMPLETED)
        
        return loss.item()

    trainer = Engine(update)
    trainer.register_events(*BackpropEvents)

    @trainer.on(BackpropEvents.BACKWARD_STARTED)
    def function_before_backprop(engine):
        # insert custom function here

```

**Comment by anmolsjoshi** [comment]
@vfdev-5 agree with your suggestion. I'll incorporate and merge this PR. all ok?

**Comment by vfdev-5** [comment]
@anmolsjoshi thanks ! Yes, go ahead !

**Pull request merged**